Document built in Timetables (#23099)
mistercrunch/Airflow: docs/apache-airflow/concepts/timetable.rst

As such, Airflow allows for custom timetables to be written in plugins and used by
DAGs. An example demonstrating a custom timetable can be found in the
:doc:`/howto/timetable` how-to guide.

Built In Timetables

> -------------------

Reviewer: ```suggestion
-------------------

```

Airflow comes with several common timetables built in to cover the most common use cases. Additional timetables
may be available in provider packages.

CronDataIntervalTimetable

> -------------------------

Reviewer: ```suggestion
^^^^^^^^^^^^^^^^^^^^^^^^^

```

> Set timing based on a cron schedule. Can be selected by providing a string that is a valid

Reviewer: ```suggestion
Set schedule based on a cron expression. Can be selected by providing a string that is a valid
```
I think this sounds better?

>   cron expression to the schedule_interval parameter of a DAG as described in the :doc:`/concepts/dags` documentation.

Reviewer: ```suggestion
  cron expression to the ``schedule_interval`` parameter of a DAG as described in the :doc:`/concepts/dags` documentation.
```

.. code-block:: python
    @dag(
        schedule_interval="0 1 * * 3"  # At 01:00 on Wednesday.
        ...
    )

DeltaDataIntervalTimetable

> --------------------------

Reviewer: ```suggestion
^^^^^^^^^^^^^^^^^^^^^^^^^^

```

Schedules data intervals with a time delta. Can be selected by providing a

>   ``datetime.timedelta`` or ``dateutil.relativedelta.relativedelta`` to the schedule_interval parameter of a DAG.

Reviewer: ```suggestion
  ``datetime.timedelta`` or ``dateutil.relativedelta.relativedelta`` to the ``schedule_interval`` parameter of a DAG.
```

.. code-block:: python
    @dag(

>         schedule_interval=datetime.timedelta(minutes=30)  # At 01:00 on Wednesday.

Reviewer: ```suggestion
        schedule_interval=datetime.timedelta(minutes=30)
```

        ...
    )

EventsTimetable

> ---------------

Reviewer: ```suggestion
^^^^^^^^^^^^^^^

```

> Simply pass a list of datetimes for the DAG to run after. Useful for timing based on sporting

Reviewer: ```suggestion
Simply pass a list of ``datetimes`` for the DAG to run after. Useful for timing based on sporting
```

As such, Airflow allows for custom timetables to be written in plugins and used by
DAGs. An example demonstrating a custom timetable can be found in the
:doc:`/howto/timetable` how-to guide.

Built In Timetables
-------------------

Airflow comes with several common timetables built in to cover the most common use cases. Additional timetables
may be available in provider packages.

CronDataIntervalTimetable
^^^^^^^^^^^^^^^^^^^^^^^^^

Set schedule based on a cron expression. Can be selected by providing a string that is a valid
  cron expression to the ``schedule_interval`` parameter of a DAG as described in the :doc:`/concepts/dags` documentation.

> .. code-block:: python

Reviewer: ```suggestion
.. code-block:: python

```

    @dag(
        schedule_interval="0 1 * * 3"  # At 01:00 on Wednesday.
        ...
    )

DeltaDataIntervalTimetable
^^^^^^^^^^^^^^^^^^^^^^^^^^

Schedules data intervals with a time delta. Can be selected by providing a
  ``datetime.timedelta`` or ``dateutil.relativedelta.relativedelta`` to the ``schedule_interval`` parameter of a DAG.

> .. code-block:: python

Reviewer: ```suggestion
.. code-block:: python

```

    @dag(
        schedule_interval=datetime.timedelta(minutes=30)
        ...
    )

EventsTimetable
^^^^^^^^^^^^^^^

Simply pass a list of ``datetimes`` for the DAG to run after. Useful for timing based on sporting
events, planned communication campaigns, and other schedules that are arbitrary and irregular but predictable.

The list of events must be finite and of reasonable size as it must be loaded every time the DAG is parsed. Optionally,
the ``restrict_to_events`` flag can be used to force manual runs of the DAG to use the time of the most recent (or very
first) event for the data interval, otherwise manual runs will run with a ``data_interval_start`` and
``data_interval_end`` equal to the time at which the manual run was begun. You can also name the set of events using the
``description`` parameter, which will be displayed in the Airflow UI.

> .. code-block:: python

Reviewer: ```suggestion
.. code-block:: python

```

As such, Airflow allows for custom timetables to be written in plugins and used by
DAGs. An example demonstrating a custom timetable can be found in the
:doc:`/howto/timetable` how-to guide.

Built In Timetables
-------------------

Airflow comes with several common timetables built in to cover the most common use cases. Additional timetables

> may be available in provider packages.

Reviewer: ```suggestion
may be available in plugins.
```

They come from plugins, not providers, right?

Author: Yes, that's true. I will correct it.

CronDataIntervalTimetable
^^^^^^^^^^^^^^^^^^^^^^^^^

Set schedule based on a cron expression. Can be selected by providing a string that is a valid
  cron expression to the ``schedule_interval`` parameter of a DAG as described in the :doc:`/concepts/dags` documentation.

.. code-block:: python

    @dag(
        schedule_interval="0 1 * * 3"  # At 01:00 on Wednesday.
        ...
    )

DeltaDataIntervalTimetable
^^^^^^^^^^^^^^^^^^^^^^^^^^

Schedules data intervals with a time delta. Can be selected by providing a

>   ``datetime.timedelta`` or ``dateutil.relativedelta.relativedelta`` to the ``schedule_interval`` parameter of a DAG.

Reviewer 2: We can use `` :class:`datetime.timedelta` `` since we already intersphinx set up for CPython. We should probably do the same for dateutil as well (maybe in a separate PR).

.. code-block:: python

    @dag(
        schedule_interval=datetime.timedelta(minutes=30)
        ...
    )

EventsTimetable
^^^^^^^^^^^^^^^

> Simply pass a list of ``datetimes`` for the DAG to run after. Useful for timing based on sporting

Reviewer 2: ```suggestion
Simply pass a list of ``datetime``s for the DAG to run after. Useful for timing based on sporting
```

events, planned communication campaigns, and other schedules that are arbitrary and irregular but predictable.

The list of events must be finite and of reasonable size as it must be loaded every time the DAG is parsed. Optionally,
the ``restrict_to_events`` flag can be used to force manual runs of the DAG to use the time of the most recent (or very
first) event for the data interval, otherwise manual runs will run with a ``data_interval_start`` and
``data_interval_end`` equal to the time at which the manual run was begun. You can also name the set of events using the
``description`` parameter, which will be displayed in the Airflow UI.

.. code-block:: python

>     @dag(
>         timetable=EventsTimetable(event_dates=[pendulum.datetime(2022, 4, 5, 8, 27,tz="America/Chicago"),
>                                                pendulum.datetime(2022, 4, 17, 8, 27,tz="America/Chicago"),
>                                                pendulum.datetime(2022, 4, 22, 20, 50,tz="America/Chicago")],
>                                   description="My Team's Baseball Games",
>                                   restrict_to_events=False)
>         ...
>     )

Reviewer 2: Let’s reformat this according to Black rules.

```suggestion
    @dag(
        timetable=EventsTimetable(
            event_dates=[
                pendulum.datetime(2022, 4, 5, 8, 27,tz="America/Chicago"),
                pendulum.datetime(2022, 4, 17, 8, 27,tz="America/Chicago"),
                pendulum.datetime(2022, 4, 22, 20, 50,tz="America/Chicago"),
            ],
            description="My Team's Baseball Games",
            restrict_to_events=False,
        ),
        ...,
    )
```

As such, Airflow allows for custom timetables to be written in plugins and used by
DAGs. An example demonstrating a custom timetable can be found in the
:doc:`/howto/timetable` how-to guide.

Built In Timetables
-------------------

Airflow comes with several common timetables built in to cover the most common use cases. Additional timetables
may be available in plugins.

CronDataIntervalTimetable
^^^^^^^^^^^^^^^^^^^^^^^^^

Set schedule based on a cron expression. Can be selected by providing a string that is a valid
  cron expression to the ``schedule_interval`` parameter of a DAG as described in the :doc:`/concepts/dags` documentation.

.. code-block:: python

    @dag(
        schedule_interval="0 1 * * 3"  # At 01:00 on Wednesday.
        ...
    )

DeltaDataIntervalTimetable
^^^^^^^^^^^^^^^^^^^^^^^^^^

Schedules data intervals with a time delta. Can be selected by providing a

>   `:class:`datetime.timedelta` or ``dateutil.relativedelta.relativedelta`` to the ``schedule_interval`` parameter of a DAG.

Reviewer 2: ```suggestion
  :class:`datetime.timedelta` or ``dateutil.relativedelta.relativedelta`` to the ``schedule_interval`` parameter of a DAG.
```

.. code-block:: python

    @dag(
        schedule_interval=datetime.timedelta(minutes=30)
        ...
    )

EventsTimetable
^^^^^^^^^^^^^^^

Simply pass a list of ``datetime``s for the DAG to run after. Useful for timing based on sporting
events, planned communication campaigns, and other schedules that are arbitrary and irregular but predictable.

The list of events must be finite and of reasonable size as it must be loaded every time the DAG is parsed. Optionally,
the ``restrict_to_events`` flag can be used to force manual runs of the DAG to use the time of the most recent (or very
first) event for the data interval, otherwise manual runs will run with a ``data_interval_start`` and
``data_interval_end`` equal to the time at which the manual run was begun. You can also name the set of events using the
``description`` parameter, which will be displayed in the Airflow UI.

.. code-block:: python

>     @dag(

Reviewer: We might want to show the import for `EventsTimetable` as well.

As such, Airflow allows for custom timetables to be written in plugins and used by
DAGs. An example demonstrating a custom timetable can be found in the
:doc:`/howto/timetable` how-to guide.

Built In Timetables
-------------------

Airflow comes with several common timetables built in to cover the most common use cases. Additional timetables
may be available in plugins.

CronDataIntervalTimetable
^^^^^^^^^^^^^^^^^^^^^^^^^

Set schedule based on a cron expression. Can be selected by providing a string that is a valid

>   cron expression to the ``schedule_interval`` parameter of a DAG as described in the :doc:`/concepts/dags` documentation.

Reviewer: ```suggestion
cron expression to the ``schedule_interval`` parameter of a DAG as described in the :doc:`/concepts/dags` documentation.
```

.. code-block:: python

    @dag(
        schedule_interval="0 1 * * 3"  # At 01:00 on Wednesday.

>         ...

Reviewer: ```suggestion
```

These are what's blowing up the static checks.

    )

DeltaDataIntervalTimetable
^^^^^^^^^^^^^^^^^^^^^^^^^^

Schedules data intervals with a time delta. Can be selected by providing a

>   :class:`datetime.timedelta` or ``dateutil.relativedelta.relativedelta`` to the ``schedule_interval`` parameter of a DAG.

Reviewer: ```suggestion
:class:`datetime.timedelta` or ``dateutil.relativedelta.relativedelta`` to the ``schedule_interval`` parameter of a DAG.
```

.. code-block:: python

    @dag(
        schedule_interval=datetime.timedelta(minutes=30)

>         ...

Reviewer: ```suggestion
```

    )

EventsTimetable
^^^^^^^^^^^^^^^

> Simply pass a list of ``datetime``s for the DAG to run after. Useful for timing based on sporting

Reviewer: ```suggestion
Simply pass a list of ``datetime``\s for the DAG to run after. Useful for timing based on sporting
```

And this should fix the docs error?

events, planned communication campaigns, and other schedules that are arbitrary and irregular but predictable.

The list of events must be finite and of reasonable size as it must be loaded every time the DAG is parsed. Optionally,
the ``restrict_to_events`` flag can be used to force manual runs of the DAG to use the time of the most recent (or very
first) event for the data interval, otherwise manual runs will run with a ``data_interval_start`` and
``data_interval_end`` equal to the time at which the manual run was begun. You can also name the set of events using the
``description`` parameter, which will be displayed in the Airflow UI.

.. code-block:: python

    from airflow.timetables.events import EventsTimetable

    @dag(
        timetable=EventsTimetable(
            event_dates=[
                pendulum.datetime(2022, 4, 5, 8, 27,tz="America/Chicago"),
                pendulum.datetime(2022, 4, 17, 8, 27,tz="America/Chicago"),
                pendulum.datetime(2022, 4, 22, 20, 50,tz="America/Chicago"),
            ],
            description="My Team's Baseball Games",
            restrict_to_events=False,
        ),

>         ...,

Reviewer: ```suggestion
```
